File: sections/67_persistence_rollback_time/problems/c_dynamic_connectivity/solution.cpp
#include <bits/stdc++.h>
using namespace std;
struct DSU {
  vector<int> parent, size;
  vector<pair<int, int>> history;
  DSU(int n) : parent(n), size(n, 1) { iota(parent.begin(), parent.end(), 0); }
  int find(int x) {
    while (x != parent[x]) x = parent[x];
    return x;
  }
  int snapshot() { return history.size(); }
  void join(int a, int b) {
    a = find(a);
    b = find(b);
    if (a == b) return;
    if (size[a] < size[b]) swap(a, b);
    history.push_back({b, size[a]});
    parent[b] = a;
    size[a] += size[b];
  }
  void rollback(int snap) {
    while ((int)history.size() > snap) {
      auto [b, old] = history.back();
      history.pop_back();
      int a = parent[b];
      size[a] = old;
      parent[b] = b;
    }
  }
};
int main() {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);
  int n, q;
  if (!(cin >> n >> q)) return 0;
  struct Op {
    char type;
    int u, v;
  };
  vector<Op> ops(q);
  map<pair<int, int>, int> opened;
  vector<vector<pair<int, int>>> seg(4 * q + 4);
  function<void(int, int, int, int, int, pair<int, int>)> add =
      [&](int node, int lo, int hi, int ql, int qr, pair<int, int> e) {
        if (ql >= hi || qr <= lo) return;
        if (ql <= lo && hi <= qr) {
          seg[node].push_back(e);
          return;
        }
        int mid = (lo + hi) / 2;
        add(node * 2, lo, mid, ql, qr, e);
        add(node * 2 + 1, mid, hi, ql, qr, e);
      };
  for (int t = 0; t < q; t++) {
    cin >> ops[t].type >> ops[t].u >> ops[t].v;
    --ops[t].u;
    --ops[t].v;
    if (ops[t].u > ops[t].v) swap(ops[t].u, ops[t].v);
    auto e = pair{ops[t].u, ops[t].v};
    if (ops[t].type == '+')
      opened[e] = t;
    else if (ops[t].type == '-') {
      add(1, 0, q, opened[e], t, e);
      opened.erase(e);
    }
  }
  for (auto [e, start] : opened) add(1, 0, q, start, q, e);
  DSU dsu(n);
  function<void(int, int, int)> dfs = [&](int node, int lo, int hi) {
    int snap = dsu.snapshot();
    for (auto [u, v] : seg[node]) dsu.join(u, v);
    if (hi - lo == 1) {
      if (ops[lo].type == '?')
        cout << (dsu.find(ops[lo].u) == dsu.find(ops[lo].v) ? "YES\n" : "NO\n");
    } else {
      int mid = (lo + hi) / 2;
      dfs(node * 2, lo, mid);
      dfs(node * 2 + 1, mid, hi);
    }
    dsu.rollback(snap);
  };
  if (q) dfs(1, 0, q);
}

File: sections/67_persistence_rollback_time/problems/c_dynamic_connectivity/solution.py
import sys


def main() -> None:
    tokens = sys.stdin.buffer.read().split()
    if not tokens:
        return
    n, q = map(int, tokens[:2])
    idx = 2
    ops = []
    opened = {}
    seg = [[] for _ in range(4 * q + 4)]

    def add(node, lo, hi, ql, qr, edge):
        if ql >= hi or qr <= lo:
            return
        if ql <= lo and hi <= qr:
            seg[node].append(edge)
            return
        mid = (lo + hi) // 2
        add(node * 2, lo, mid, ql, qr, edge)
        add(node * 2 + 1, mid, hi, ql, qr, edge)

    for time in range(q):
        kind = tokens[idx].decode()
        u = int(tokens[idx + 1]) - 1
        v = int(tokens[idx + 2]) - 1
        idx += 3
        if u > v:
            u, v = v, u
        ops.append((kind, u, v))
        edge = (u, v)
        if kind == "+":
            opened[edge] = time
        elif kind == "-":
            add(1, 0, q, opened.pop(edge), time, edge)
    for edge, start in opened.items():
        add(1, 0, q, start, q, edge)
    parent = list(range(n))
    size = [1] * n
    history = []

    def find(x):
        while x != parent[x]:
            x = parent[x]
        return x

    def join(a, b):
        a = find(a)
        b = find(b)
        if a == b:
            return
        if size[a] < size[b]:
            a, b = b, a
        history.append((b, size[a]))
        parent[b] = a
        size[a] += size[b]

    def rollback(snapshot):
        while len(history) > snapshot:
            b, old = history.pop()
            a = parent[b]
            size[a] = old
            parent[b] = b

    output = []

    def dfs(node, lo, hi):
        snapshot = len(history)
        for u, v in seg[node]:
            join(u, v)
        if hi - lo == 1:
            kind, u, v = ops[lo]
            if kind == "?":
                output.append("YES" if find(u) == find(v) else "NO")
        else:
            mid = (lo + hi) // 2
            dfs(node * 2, lo, mid)
            dfs(node * 2 + 1, mid, hi)
        rollback(snapshot)

    if q:
        dfs(1, 0, q)
    print("\n".join(output))


if __name__ == "__main__":
    main()
